### PR TITLE
Enable XLC16.1 toolchain for AIX platforms

### PR DESCRIPTION
**Pull request by midronij**
Added -lc++ library to linker as a temporary fix. Since the linker command being used, `makeC++SharedLib_r` is used to link legacy object files that don't depend on -lc++, it does not automatically add this library, so it needs to be added manually. 

Related to https://github.com/eclipse/omr/pull/4285

Signed-off-by: Jackie Midroni <jackie.midroni@mail.utoronto.ca>

### CODE REVIEW

File: runtime/makelib/targets.mk.aix.inc.ftl [hunk 1]
```diff
@@ -158,7 +158,7 @@ $(patsubst %.s,%.o,$(filter %.s,$(UMA_FILES_TO_PREPROCESS))) : %$(UMA_DOT_O) : %
 
 ifdef UMA_TREAT_WARNINGS_AS_ERRORS
   ifndef UMA_SUPPRESS_WARNINGS_AS_ERRORS
-    CFLAGS += -qhalt=w
-    CXXFLAGS += -qhalt=w
+    CFLAGS += -qhalt=e
```
Comment (pshipton) [inline runtime/makelib/targets.mk.aix.inc.ftl]: What does this fix? Although you said `likely not feasible as long term solutions`, we don't want this change even as a short term solution as it's hard to get back to a state with no warnings once warning are allowed to creep in.

Comment (pshipton) [reply runtime/makelib/targets.mk.aix.inc.ftl]: A better way to do this, although still better to fix the warnings, is to disable UMA_TREAT_WARNINGS_AS_ERRORS for only the subdirectory that needs it.

Comment (midronij) [reply runtime/makelib/targets.mk.aix.inc.ftl]: This was intended to fix this error: 
```16:13:46  jnitest.c:1020:1: error: control reaches end of non-void function [-Werror,-Wreturn-type]```
At that point my priority was getting the build to succeed so that we could compare performance between xlC and xlclang, but since that's no longer the case, should I just scrap this change entirely?

Comment (pshipton) [reply runtime/makelib/targets.mk.aix.inc.ftl]: Yes, please remove this part of the change. You can do this for your personal builds, but we should fix the jnitest.c warning properly. Which can be done in another PR.

File: runtime/tests/jni/module.xml [hunk 2]
```diff
@@ -378,7 +378,7 @@
 			<include path="j9oti"/>
 		</includes>
 		<makefilestubs>
-			<makefilestub data="UMA_TREAT_WARNINGS_AS_ERRORS=1"/>
+			<makefilestub data="UMA_TREAT_WARNINGS_AS_ERRORS=0"/>
```
Comment (pshipton) [inline runtime/tests/jni/module.xml]: Now I understand why this was added, I can say we don't want to deliver even this. Either fix the code in jnitest.c to remove the warning, or leave it for somebody else to fix.

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk [hunk 3]
```diff
@@ -193,13 +193,13 @@ IPP_FLAGS+=$(IPP_FLAGS_EXTRA)
 #
 SOLINK_CMD?=$(SHAREDLIB)
 
-SOLINK_FLAGS+=-p0 -bloadmap:lmap -brtl -bnoentry -bnolibpath
+SOLINK_FLAGS+=-p0 -bloadmap:lmap -brtl -bnoentry -bnolibpath
```
Comment (pshipton) [inline runtime/compiler/build/toolcfg/aix-xlc/common.mk]: I don't expect a space needs to be added here.

File: runtime/tests/jni/module.xml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
 <!--
-   Copyright (c) 2006, 2018 IBM Corp. and others
+   Copyright (c) 2006, 2019 IBM Corp. and others
```
Comment (pshipton) [inline runtime/tests/jni/module.xml]: Pls remove this unnecessary change.

**Comment by pshipton** [comment]
Pls update the commit comment to explain why this change is needed.

File: runtime/tests/jni/module.xml (hunk 4 above)
Comment (midronij) [reply runtime/tests/jni/module.xml]: Without this change, the copyright check fails. Is there a better way to prevent that from happening?

Comment (pshipton) [reply runtime/tests/jni/module.xml]: It won't fail if you remove this change, because you've removed the other change in this file as well.

**Comment by pshipton** [comment]
jenkins compile aix jdk8,jdk13

**Comment by pshipton** [comment]
@midronij please create a PR with the same change against the v0.16.0-release branch,

**Comment by pshipton** [comment]
This is causing an AIX build problem, I'm going to revert it while we figure it out.
```
ld: 0706-006 Cannot find or open library file: -l c++
        ld:open(): A file or directory in the path name does not exist.
```

**Comment by midronij** [comment]
@pshipton Here's the 0.16.0 PR: https://github.com/eclipse/openj9/pull/7047

**Comment by pshipton** [comment]
@midronij are you working on a better fix that doesn't cause the `Cannot find or open library file: -l c++` problem?
@zl-wang

**Comment by zl-wang** [comment]
the existing build seemed breaking loose everything on AIX.  dump utility is not working ... dbx crashed itself with internal error ... symbol names cannot be demangled without new libC.a on system.  The climax is the interpreter performance ... details below.

This just looks like a mere couple 100x slower than expected. @pshipton @DanHeidinga is this a strong enough justification to go back to xlC build (using xlC16 is fine)? I understand the investigation is preferred to be completed.  I actually suspect if the existing build is a debug build (so that it is slow).  Anyway, the facts:

milan:[j9.64] time bin/java -Xint -version
openjdk version "13-internal" 2019-09-17
OpenJDK Runtime Environment (build 13-internal+0-adhoc.jenkins.openj9-openjdk-jdk13)
Eclipse OpenJ9 VM (build aixxlc16-2f6383c98, JRE 13 AIX ppc64-64-Bit Compressed References 20190923_000000 (JIT disabled, AOT disabled)
OpenJ9   - 2f6383c98
OMR      - 69029bae
JCL      - b4b82f558d based on jdk-13+33)
11.874u 0.483s 0:21.44 57.6%    3+224k 0+0io 67pf+0w

milan:[j9.64] time bin/java -Xint -version
openjdk version "13-internal" 2019-09-17
OpenJDK Runtime Environment (build 13-internal+0-adhoc.jenkins.openj9-openjdk-jdk13)
Eclipse OpenJ9 VM (build aixxlc16-2f6383c98, JRE 13 AIX ppc64-64-Bit Compressed References 20190923_000000 (JIT disabled, AOT disabled)
OpenJ9   - 2f6383c98
OMR      - 69029bae
JCL      - b4b82f558d based on jdk-13+33)
11.872u 0.490s 0:21.36 57.8%    3+224k 0+0io 8pf+0w

milan:[j9.64] time bin/java -Xint -version
openjdk version "13-internal" 2019-09-17
OpenJDK Runtime Environment (build 13-internal+0-adhoc.jenkins.openj9-openjdk-jdk13)
Eclipse OpenJ9 VM (build aixxlc16-2f6383c98, JRE 13 AIX ppc64-64-Bit Compressed References 20190923_000000 (JIT disabled, AOT disabled)
OpenJ9   - 2f6383c98
OMR      - 69029bae
JCL      - b4b82f558d based on jdk-13+33)
11.872u 0.503s 0:21.33 57.9%    3+222k 0+0io 8pf+0w

**Comment by DanHeidinga** [comment]
@zl-wang My understanding is the OpenJDK community has switched to xlc16 / xlclang++.  We aim to use the same compile environment as OpenJDK to ensure compatibility between the class library natives and our compile environment.  Part of the reason for their switch is http://openjdk.java.net/jeps/347 which allows for using C++14 features in the OpenJDK codebase.  

> xlclang++, to invoke the new Clang-based XL C++ compiler that adheres to the C++11 and C++14 standards

From http://www-01.ibm.com/common/ssi/ShowDoc.wss?docURL=/common/ssi/rep_ca/4/877/ENUSZP18-0674/index.html&request_locale=en , it appears C++14 is only available with xlclang++ frontend.  

I'm not sure that leaves us any choice other than to work through the issues on xlclang++ and ensure OpenJ9 continues to perform well in this configuration.

**Comment by gita-omr** [comment]
Alright, we will look for the check we need to use.

**Comment by midronij** [comment]
I created a new pull request with the suggested changes here: https://github.com/eclipse/openj9/pull/7499
